{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ingest gene expression data from TCGA, TARGET and GTEX\n",
    "\n",
    "Download gene expression data from the UCSC Xena Toil re-compute dataset. This dataset comprises gene expression data for 19 thousand samples all processed using the same genomics pipeline and therefore comparable to each other. Each sample is a vector of the floating point expression relative level for each of 60k genes. Associated with these data is clinical information on each sample such as type (tumor vs. normal), primary site (where the sample came from) etc... We use this information to split the dataset into one-hot tumor vs. normal as well as into a set of classes based on primary site for stratification.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 8.99 s, sys: 46 s, total: 55 s\n",
      "Wall time: 55.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\"\"\"\n",
    "Download the expression data from xena, if we haven't already, and save in \n",
    "an hdf5 file. This can take around 30 minutes between the download and\n",
    "the conversion from tsv into float32 dataframes so we save the tsv and hdf5\n",
    "file for future use. We download manually vs. passing read_csv a url \n",
    "directly as the latter times out with this size file. Note we invert\n",
    "the expression matrix to conform to machine learning where the samples\n",
    "are in rows vs. the standard genomics world where samples columns\n",
    "and reatures (i.e. genes) are rows.\n",
    "\"\"\"\n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm\"):\n",
    "    print(\"Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\")\n",
    "    with open(\"data/TcgaTargetGtex_rsem_gene_tpm\", \"wb\") as f:\n",
    "        r = requests.get(\"https://toil.xenahubs.net/download/TcgaTargetGtex_rsem_gene_tpm\", stream=True)\n",
    "        r.raise_for_status()\n",
    "        for block in r.iter_content(32768):\n",
    "            f.write(block)\n",
    "\n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    expression = pd.read_csv(\"data/TcgaTargetGtex_rsem_gene_tpm\", \n",
    "                             sep=\"\\t\", index_col=0).dropna().astype(np.float32).T\n",
    "    expression.to_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "X = pd.read_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\").sort_index(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>sample</th>\n",
       "      <th>ENSG00000242268.2</th>\n",
       "      <th>ENSG00000259041.1</th>\n",
       "      <th>ENSG00000270112.3</th>\n",
       "      <th>ENSG00000167578.16</th>\n",
       "      <th>ENSG00000278814.1</th>\n",
       "      <th>ENSG00000078237.5</th>\n",
       "      <th>ENSG00000269416.5</th>\n",
       "      <th>ENSG00000263642.1</th>\n",
       "      <th>ENSG00000146083.11</th>\n",
       "      <th>ENSG00000158486.13</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000009694.13</th>\n",
       "      <th>ENSG00000238244.3</th>\n",
       "      <th>ENSG00000216352.1</th>\n",
       "      <th>ENSG00000123685.8</th>\n",
       "      <th>ENSG00000267117.1</th>\n",
       "      <th>ENSG00000273233.1</th>\n",
       "      <th>ENSG00000105063.18</th>\n",
       "      <th>ENSG00000231119.2</th>\n",
       "      <th>ENSG00000280861.1</th>\n",
       "      <th>ENSG00000181518.3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.2934</td>\n",
       "      <td>5.1190</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.8488</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>5.1498</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>2.4571</td>\n",
       "      <td>-1.4305</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>2.8178</td>\n",
       "      <td>-0.1031</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>5.1631</td>\n",
       "      <td>-3.3076</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>4.1277</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.6880</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.4830</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.9132</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.9406</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-1.5105</td>\n",
       "      <td>4.1764</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>4.4067</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.0440</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>4.3841</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>1.5165</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>1.7141</td>\n",
       "      <td>-1.1488</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>4.8768</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>-1.2481</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>5.6860</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>1.3679</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>5.0644</td>\n",
       "      <td>-1.8836</td>\n",
       "      <td>...</td>\n",
       "      <td>1.5998</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.9356</td>\n",
       "      <td>-1.1488</td>\n",
       "      <td>-1.0559</td>\n",
       "      <td>4.8694</td>\n",
       "      <td>-1.9379</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>-3.8160</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.3573</td>\n",
       "      <td>4.0357</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.4325</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.9421</td>\n",
       "      <td>-3.4580</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.1714</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.6608</td>\n",
       "      <td>-1.1811</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>3.6816</td>\n",
       "      <td>-2.6349</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 60498 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "sample                    ENSG00000242268.2  ENSG00000259041.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -1.2481            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -3.8160            -9.9658   \n",
       "\n",
       "sample                    ENSG00000270112.3  ENSG00000167578.16  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -4.2934              5.1190   \n",
       "GTEX-1117F-0426-SM-5EGHI             0.0014              4.1277   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658              4.4067   \n",
       "GTEX-1117F-0626-SM-5N9CS            -5.5735              5.6860   \n",
       "GTEX-1117F-0726-SM-5GIEN             0.3573              4.0357   \n",
       "\n",
       "sample                    ENSG00000278814.1  ENSG00000078237.5  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658             0.8488   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658             0.6880   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658             0.0440   \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658             1.3679   \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658            -0.4325   \n",
       "\n",
       "sample                    ENSG00000269416.5  ENSG00000263642.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -5.5735            -9.9658   \n",
       "\n",
       "sample                    ENSG00000146083.11  ENSG00000158486.13  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7              5.1498             -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI              3.4830             -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ              4.3841             -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS              5.0644             -1.8836   \n",
       "GTEX-1117F-0726-SM-5GIEN              3.9421             -3.4580   \n",
       "\n",
       "sample                          ...          ENSG00000009694.13  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7        ...                      2.4571   \n",
       "GTEX-1117F-0426-SM-5EGHI        ...                     -0.9132   \n",
       "GTEX-1117F-0526-SM-5EGHJ        ...                      1.5165   \n",
       "GTEX-1117F-0626-SM-5N9CS        ...                      1.5998   \n",
       "GTEX-1117F-0726-SM-5GIEN        ...                     -3.1714   \n",
       "\n",
       "sample                    ENSG00000238244.3  ENSG00000216352.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -1.4305            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658            -9.9658   \n",
       "\n",
       "sample                    ENSG00000123685.8  ENSG00000267117.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7             2.8178            -0.1031   \n",
       "GTEX-1117F-0426-SM-5EGHI            -0.9406            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ             1.7141            -1.1488   \n",
       "GTEX-1117F-0626-SM-5N9CS             3.9356            -1.1488   \n",
       "GTEX-1117F-0726-SM-5GIEN             0.6608            -1.1811   \n",
       "\n",
       "sample                    ENSG00000273233.1  ENSG00000105063.18  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658              5.1631   \n",
       "GTEX-1117F-0426-SM-5EGHI            -1.5105              4.1764   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658              4.8768   \n",
       "GTEX-1117F-0626-SM-5N9CS            -1.0559              4.8694   \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658              3.6816   \n",
       "\n",
       "sample                    ENSG00000231119.2  ENSG00000280861.1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7            -3.3076            -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI            -5.0116            -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658            -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS            -1.9379            -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN            -2.6349            -9.9658   \n",
       "\n",
       "sample                    ENSG00000181518.3  \n",
       "GTEX-1117F-0226-SM-5GZZ7            -9.9658  \n",
       "GTEX-1117F-0426-SM-5EGHI            -9.9658  \n",
       "GTEX-1117F-0526-SM-5EGHJ            -9.9658  \n",
       "GTEX-1117F-0626-SM-5N9CS            -9.9658  \n",
       "GTEX-1117F-0726-SM-5GIEN            -9.9658  \n",
       "\n",
       "[5 rows x 60498 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>detailed_category</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>Muscle - Skeletal</td>\n",
       "      <td>Muscle</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>Artery - Tibial</td>\n",
       "      <td>Blood Vessel</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>Artery - Coronary</td>\n",
       "      <td>Blood Vessel</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>Heart - Atrial Appendage</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 detailed_category    primary_site  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7    Adipose - Subcutaneous  Adipose Tissue   \n",
       "GTEX-1117F-0426-SM-5EGHI         Muscle - Skeletal          Muscle   \n",
       "GTEX-1117F-0526-SM-5EGHJ           Artery - Tibial    Blood Vessel   \n",
       "GTEX-1117F-0626-SM-5N9CS         Artery - Coronary    Blood Vessel   \n",
       "GTEX-1117F-0726-SM-5GIEN  Heart - Atrial Appendage           Heart   \n",
       "\n",
       "                            sample_type  gender study tumor_normal  \n",
       "GTEX-1117F-0226-SM-5GZZ7  Normal Tissue  Female  GTEX       Normal  \n",
       "GTEX-1117F-0426-SM-5EGHI  Normal Tissue  Female  GTEX       Normal  \n",
       "GTEX-1117F-0526-SM-5EGHJ  Normal Tissue  Female  GTEX       Normal  \n",
       "GTEX-1117F-0626-SM-5N9CS  Normal Tissue  Female  GTEX       Normal  \n",
       "GTEX-1117F-0726-SM-5GIEN  Normal Tissue  Female  GTEX       Normal  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the clinical (i.e. labels) from xena\n",
    "Y = pd.read_table(\"https://toil.xenahubs.net/download/TcgaTargetGTEX_phenotype.txt\",\n",
    "                  header=0, names=[\"detailed_category\", \"primary_site\", \"sample_type\", \"gender\", \"study\"],\n",
    "                  sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\").sort_index(axis=0)\n",
    "\n",
    "# Compute and add a tumor/normal column\n",
    "Y[\"tumor_normal\"] = Y.apply(\n",
    "    lambda row: \"Normal\" if row[\"sample_type\"] in [\"Cell Line\", \"Normal Tissue\", \"Solid Tissue Normal\"]\n",
    "    else \"Tumor\", axis=1)\n",
    "\n",
    "Y.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>detailed_category</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>19130</td>\n",
       "      <td>19126</td>\n",
       "      <td>19131</td>\n",
       "      <td>18972</td>\n",
       "      <td>19131</td>\n",
       "      <td>19131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>93</td>\n",
       "      <td>46</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1212</td>\n",
       "      <td>1846</td>\n",
       "      <td>9185</td>\n",
       "      <td>10456</td>\n",
       "      <td>10535</td>\n",
       "      <td>10531</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                detailed_category primary_site    sample_type gender  study  \\\n",
       "count                       19130        19126          19131  18972  19131   \n",
       "unique                         93           46             17      2      3   \n",
       "top     Breast Invasive Carcinoma        Brain  Primary Tumor   Male   TCGA   \n",
       "freq                         1212         1846           9185  10456  10535   \n",
       "\n",
       "       tumor_normal  \n",
       "count         19131  \n",
       "unique            2  \n",
       "top           Tumor  \n",
       "freq          10531  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19126 samples with non-null labels\n"
     ]
    }
   ],
   "source": [
    "# Remove rows where the label is null or the sample is missing\n",
    "Y_not_null = Y[pd.notnull(Y[\"primary_site\"])]\n",
    "intersection = X.index.intersection(Y_not_null.index)\n",
    "X_clean = X[X.index.isin(intersection)]\n",
    "Y_clean = Y[Y.index.isin(intersection)]\n",
    "\n",
    "# Make sure the label and example samples are in the same order\n",
    "assert(X_clean.index.equals(Y_clean.index))\n",
    "\n",
    "print(intersection.shape[0], \"samples with non-null labels\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Convert tumor/normal labels to binary 1/0\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "encoder = LabelEncoder()\n",
    "y_binary = encoder.fit_transform(Y_clean[\"tumor_normal\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 16.1 s, sys: 11 s, total: 27.1 s\n",
      "Wall time: 26.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Split into stratified training and test sets based on classes (i.e. tissue type)\n",
    "from sklearn.model_selection import StratifiedShuffleSplit\n",
    "split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "for train_index, test_index in split.split(X_clean.values, Y_clean[\"primary_site\"]):\n",
    "    X_train, X_test = X.values[train_index], X_clean.values[test_index]\n",
    "    y_train, y_test = y_binary[train_index], y_binary[test_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Convert from ensembl to hugo gene names as the latter are more commonly known\n",
    "# ensembl = pd.read_table(\"ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt\", \n",
    "#                         usecols=[\"symbol\", \"ensembl_gene_id\", \"gene_family\"],\n",
    "#                         index_col=2)\n",
    "# ensembl.head()\n",
    "# ensembl.loc[\"ENSG00000121410\"].symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Write to an h5 file for training\n",
    "X: examples as a numpy floating point array\n",
    "y: binary to predict\n",
    "features: string array of the features in X (Ensemble gene ids)\n",
    "labels: array of strings coresponding to the binary labels\n",
    "\"\"\"\n",
    "with h5py.File(\"data/tumor_normal.h5\", \"w\") as f:\n",
    "    f.create_dataset('X_train', X_train.shape, dtype='f')[:] = X_train\n",
    "    f.create_dataset('X_test', X_test.shape, dtype='f')[:] = X_test\n",
    "    f.create_dataset('y_train', y_train.shape, dtype='i')[:] = y_train\n",
    "    f.create_dataset('y_test', y_test.shape, dtype='i')[:] = y_test\n",
    "    f.create_dataset('features', X_clean.columns.shape, 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in X_clean.columns.values])\n",
    "    f.create_dataset('labels', (2, 1), 'S10', \n",
    "                     [l.encode(\"ascii\", \"ignore\") for l in [\"Normal\", \"Tumor\"]])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
